File: src/examples/Decompose.c
#include <vkt/Decompose.h>
#include <vkt/Fill.h>
#include <vkt/linalg.h>
#include <vkt/Render.h>
#include <vkt/StructuredVolume.h>

int main()
{
    // Volume dimensions
    vktVec3i_t dims = { .x=120,.y=66,.z=49 };

    // Brick size
    vktVec3i_t brickSize = { .x=16, .y=16, .z=16 };

    // Halo / ghost cells
    vktVec3i_t haloSizeNeg = { .x=1, .y=1, .z=1 };
    vktVec3i_t haloSizePos = { .x=1, .y=1, .z=1 };

    int bpv = 1;

    float mappingLo = 0.f;
    float mappingHi = 1.f;

    float distX = 1.f;
    float distY = 1.f;
    float distZ = 1.f;

    vktStructuredVolume volume;

    // The destination data structure
    vktArray3D_vktStructuredVolume decomp;

    vktStructuredVolumeCreate(
            &volume,
            dims.x,
            dims.y,
            dims.z,
            bpv,
            distX,
            distY,
            distZ,
            mappingLo,
            mappingHi
            );

    vktArray3D_vktStructuredVolume_CreateEmpty(&decomp);

    // Put some values in
    vktFillSV(volume, .1f);

    // Preallocate storage for the decomposition
    vktBrickDecomposeResizeSV(
            decomp,
            volume,
            brickSize.x,
            brickSize.y,
            brickSize.z,
            haloSizeNeg.x,
            haloSizeNeg.y,
            haloSizeNeg.z,
            haloSizePos.x,
            haloSizePos.y,
            haloSizePos.z
            );

    // Compute the decomposition
    vktBrickDecomposeSV(
            decomp,
            volume,
            brickSize.x,
            brickSize.y,
            brickSize.z,
            haloSizeNeg.x,
            haloSizeNeg.y,
            haloSizeNeg.z,
            haloSizePos.x,
            haloSizePos.y,
            haloSizePos.z
            );

    vktVec3i_t index = { .x=0, .y=0, .z=0 };
    vktRenderState_t renderState;
    vktRenderStateDefaultInit(&renderState);
    vktRenderSV(
        *vktArray3D_vktStructuredVolume_Access(decomp, index),
        renderState,
        NULL
        );

    vktArray3D_vktStructuredVolume_Destroy(decomp);
    vktStructuredVolumeDestroy(volume);
}

File: src/examples/Decompose.py
#!/usr/bin/python3

import volkit as vkt

def main():
    # Volume dimensions
    dims = vkt.Vec3i()
    dims.x = 120
    dims.y = 66
    dims.z = 49

    # Brick size
    brickSize = vkt.Vec3i()
    brickSize.x = 16
    brickSize.y = 16
    brickSize.z = 16

    # Halo / ghost cells
    haloSizeNeg = vkt.Vec3i()
    haloSizeNeg.x = 1
    haloSizeNeg.y = 1
    haloSizeNeg.z = 1
    haloSizePos = vkt.Vec3i()
    haloSizePos.x = 1
    haloSizePos.y = 1
    haloSizePos.z = 1

    dataFormat = vkt.DataFormat_UInt8

    mappingLo = 0.
    mappingHi = 1.

    distX = 1.
    distY = 1.
    distZ = 1.

    volume = vkt.StructuredVolume(
            dims.x,
            dims.y,
            dims.z,
            dataFormat,
            distX,
            distY,
            distZ,
            mappingLo,
            mappingHi
            )

    # Put some values in
    vkt.Fill(volume, .1)

    # The destination data structure
    decomp = vkt.Array3D_StructuredVolume()

    # Preallocate storage for the decomposition
    vkt.BrickDecomposeResize(
            decomp,
            volume,
            brickSize,
            haloSizeNeg,
            haloSizePos
            )

    # Compute the decomposition
    vkt.BrickDecompose(
            decomp,
            volume,
            brickSize,
            haloSizeNeg,
            haloSizePos
            )

    vkt.Render(decomp[vkt.Vec3i()])

if __name__ == '__main__':
    main()
